Adding upsampled data in csv format

Mahi7828 opened the pull request:
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
This PR addresses inconsistencies in visualizing the MOCA dataset in Napari, where only every 5th frame is tracked, leading to mismatches between the CSV data and the video 

**What does this PR do?**
- Adds data values for each frame, initially it had only for every 5th frame.
- Linear interpolation is used to generate the data which is then saved in csv format
- improved bounding boxes behavior over time for a smoother GUI experience

## References
PR #494 and PR #434 

## How has this PR been tested?
The pytest are passed. There is new file addition for upsampling MOCA dataset, the code itself has some code snippets whose outputs can be used to verify the proper interpolation of data. 

## Is this a breaking change?
No


## Does this PR require an update to the documentation?
yes

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

Mahi7828 commented:
The quality gate is failing may due to the similarity in some part of the code which I submitted in the previous PR. Please suggest modifications to be done and let me know how can I use the upsampled csv files in napari without them getting updated to gin.

sfmig commented:
Hi @Mahi7828 

Just to clarify: you cannot add data to our GIN data repository, but you can access it as a user: you can download data from the web interface and also use our `sample_data` API to fetch the datasets of interest. 

You can find a detailed example on how to use `movement` to fetch sample datasets [here](https://movement.neuroinformatics.dev/examples/load_and_explore_poses.html#sphx-glr-examples-load-and-explore-poses-py). I also recommend reading the [input/output section](https://movement.neuroinformatics.dev/user_guide/input_output.html#input-output) of the user guide.

I ran your script and generated the `upsampled_dataset.csv` file, and I was able to visualise it in `napari`. Can you tell me where you had issues loading it with napari? I followed the steps in our [GUI guide](https://movement.neuroinformatics.dev/user_guide/gui.html).

However, there seems to be a problem with the trajectory generated. The upsampled trajectory seems to track the lower right corner of the bounding box around the animal, rather than the centre. Would you mind having another look and checking why that may be? I attach a video below:

https://github.com/user-attachments/assets/fb11effb-e5d4-42e0-8f57-f0fccc1390f0


Apart from that fix, I have another suggestion:

Can you have a go at generating two upsampled trajectories, using two different methods: linear interpolation and also forward filling? You can find how to do that last one in the [same example](https://movement.neuroinformatics.dev/examples/load_and_upsample_bboxes.html#fill-in-empty-values-with-forward-filling) from which you got the linear interpolation. That way we can visually compare in the video which one does better.

Mahi7828 commented:
Hi @sfmig 
I have updated this pull request with three new code files in the examples directory. The upsampling_linear.py is for upsampling of data using linear interpolation, I have modified it such that the pointer traces the centroid of the body, next I have put code for interpolation using forward filling. Both these codes generate csv data which can be visualised on napari. 
Also, I used the code csv_napari to visualise the csv files on napari.
From my observations, linear interpolation gives better results than forward filling. 
Please let me know what more should be done for this PR.
Thankyou

sfmig commented:
Hi @Mahi7828,

Thanks for having a go. I have a few general comments you could work on.

I would suggest first to format the forward filling interpolation script as a notebook (.ipynb), and then combining that notebook with the linear upsampling one. There is a lot of overlap between both files at the moment, which is a bit unnecessary. It makes it more difficult for people to read through both files and to understand the differences between the two approaches. This can also be a good learning exercise, you can additionally try to make the notebook more user-friendly combining markdown and code cells that explain the process.

I would also recommend using our file validators to ensure the generated files are importable into `movement`. The simplest way would be to check if you can successfully create a `ValidVIAtracksCSV` object doing something like the following before exporting the csv:
```python
from movement.validators.files import ValidVIATracksCSV
from pathlib import Path
file = ValidVIATracksCSV(Path("/path/to/upsampled.csv"))
```

I am confused about the need for the `csv_napari.py` script. We already support loading VIA-tracks files in `movement` and in `napari` - as specified in the comment above, you can see how to do this in our [GUI guide](https://movement.neuroinformatics.dev/user_guide/gui.html). Our users would not need a script to do this.

I would also recommend to link the relevant issue in the PR description - this helps us keep better track of things. I noticed it is linked, but it is referred as a PR. Would you mind changing that?

Finally, I would like to highlight that this PR is an unusual one - we do not expect to merge this work since the actual contribution would be the output csv files that we will upload to our GIN data repository (with proper credit to you) when ready. Since there is already an [example](https://movement.neuroinformatics.dev/examples/load_and_upsample_bboxes.html#sphx-glr-examples-load-and-upsample-bboxes-py) that does the exact same work as here (minus the exporting step), it also doesn't make a lot of sense to expand this work into an example. But still it would be useful to fully wrap up the work, and have a reference somewhere of how these csv files were generated - even as a closed PR it will remain in the repo history. 

One thing that your work highlighted is that it is not clear from the current documentation that the x,y coordinates in the VIA tracks file refer to the [top-left corner](https://github.com/neuroinformatics-unit/movement/blob/e011ff8b4e004b950446998a9def898e18d2e3b7/movement/io/load_bboxes.py#L440) of the bounding box. This is in contrast to the `movement` dataset, where x,y coordinates in the position array refer to the centroid. It would be useful to clarify this further to our users, in the relevant function docstrings and user guides. Would you like to have a go at this? I opened an issue #536, if you want you can try opening a PR to address it.

Hope this helps!

Mahi7828 commented:
Heyy @sfmig 

I  have added a Python notebook to compare the linear and forward-filled interpolation, along with the extraction of these upsampled xarray datasets into CSV files, in a format that can be visualised in napari. Please let me know if I need to increase the depth of information currently provided or if something else needs to be improved in the IPython notebook file added. 
Also, if everything's fine, I can start working on the interpolation to create datasets for uploading on the GIN data repository. 

In my previous commit, I tried to follow the GUI Guide, but had to make the csv_napari.py file because I was facing difficulty in loading VIA-tracks directly, because there was an issue of frame size and proper orientation. I will try again to work on this loading via tracks. 

The PR has been linked in the reference section. 

The top-left corner issue, I have tried to work on it, will add an update in the PR #536. 

Please let me know what all needs to be done. 
Thankyou
